src/runtime/driver/handle.rs

```diff
@@ -0,0 +1,272 @@
+                op.index = usize::MAX;
+                Poll::Ready(op.data.take().unwrap().complete(cqe))
+            }
+            Lifecycle::CompletionList(indices) => {
+                let mut data = op.data.take().unwrap();
+                let mut status = Poll::Pending;
+                // Consume the CqeResult list, calling update on the Op on all Cqe's flagged `more`
+                // If the final Cqe is present, clean up and return Poll::Ready
+                for cqe in indices.into_list(completions) {
+                    if io_uring::cqueue::more(cqe.flags) {
+                        data.update(cqe);
+                    } else {
+                        status = Poll::Ready(cqe);
+                        break;
+                    }
+                }
+                match status {
+                    Poll::Pending => {
+                        // We need more CQE's. Restore the op state
+                        let _ = op.data.insert(data);
+                        *lifecycle = Lifecycle::Waiting(cx.waker().clone());
+                        Poll::Pending
+                    }
+                    Poll::Ready(cqe) => {
+                        driver.ops.remove(op.index);
+                        op.index = usize::MAX;
+                        Poll::Ready(data.complete(cqe))
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) fn remove_op<T, CqeType>(&self, op: &mut Op<T, CqeType>) {
+        use std::mem;
+
+        let mut driver = self.inner.borrow_mut();
+
+        // Get the Op Lifecycle state from the driver
+        let (lifecycle, completions) = match driver.ops.get_mut(op.index) {
+            Some(val) => val,
+            None => {
+                // Op dropped after the driver
+                return;
+            }
+        };
+
+        match mem::replace(lifecycle, Lifecycle::Submitted) {
+            Lifecycle::Submitted | Lifecycle::Waiting(_) => {
+                *lifecycle = Lifecycle::Ignored(Box::new(op.data.take()));
+            }
+            Lifecycle::Completed(..) => {
+                driver.ops.remove(op.index);
+            }
+            Lifecycle::CompletionList(indices) => {
+                // Deallocate list entries, recording if more CQE's are expected
+                let more = {
+                    let mut list = indices.into_list(completions);
+                    io_uring::cqueue::more(list.peek_end().unwrap().flags)
+                    // Dropping list deallocates the list entries
+                };
+                if more {
+                    // If more are expected, we have to keep the op around
+                    *lifecycle = Lifecycle::Ignored(Box::new(op.data.take()));
+                } else {
+                    driver.ops.remove(op.index);
+                }
+            }
+            Lifecycle::Ignored(..) => unreachable!(),
+        }
+    }
+}
+
+impl WeakHandle {
+    pub(crate) fn upgrade(&self) -> Option<Handle> {
```

Comment:
Output lifetime not tied to &self's lifetime. By design?

Reply:
Imagine I'll run into reasons soon enough when I looking at more code. Ignore my question.

Reply:
This retrieves a Handle from WeakHandle, and basically copies from the Rc/Arc API